front/src/modules/timesStops/helpers/computeMargins.tsx

```diff
@@ -0,0 +1,94 @@
+import type { SimulationResponseSuccess } from 'applications/operationalStudies/types';
+import type { TrainScheduleResult } from 'common/api/generatedEditoastApi';
+import type { OperationalPointWithTimeAndSpeed } from 'modules/trainschedule/components/DriverTrainScheduleV2/types';
+import { interpolateValue } from 'modules/trainschedule/components/DriverTrainScheduleV2/utils';
+import { mToMm, msToS } from 'utils/physics';
+
+import { formatDigitsAndUnit } from './utils';
+
+const emptyMargins = {
+  theoreticalMargin: '',
+  theoreticalMarginSeconds: '',
+  calculatedMarginSeconds: '',
+  diffMarginsSeconds: '',
+};
+
+function getTheoreticalMargin(selectedTrainSchedule: TrainScheduleResult, pathStepId: string) {
+  const pathStepIndex = selectedTrainSchedule.path.findIndex(
+    (pathStep) => pathStep.id === pathStepId
+  );
+  let theoreticalMarginIndex;
+  if (pathStepIndex === 0) {
+    theoreticalMarginIndex = 0;
+  } else {
+    const theoriticalMarginBoundaryIndex = selectedTrainSchedule.margins?.boundaries?.findIndex(
+      (id) => id === pathStepId
+    );
+    if (
+      theoriticalMarginBoundaryIndex === undefined ||
+      theoriticalMarginBoundaryIndex < 0 ||
+      theoriticalMarginBoundaryIndex > selectedTrainSchedule.margins!.values.length - 2
+    ) {
+      return undefined;
+    }
+    theoreticalMarginIndex = theoriticalMarginBoundaryIndex + 1;
+  }
+  return selectedTrainSchedule.margins!.values[theoreticalMarginIndex];
+}
+
+function computeDuration(
+  trainSimulation: SimulationResponseSuccess,
+  opPoint: OperationalPointWithTimeAndSpeed,
+  nextOpPoint: OperationalPointWithTimeAndSpeed,
+  key: 'base' | 'provisional' | 'final_output'
+) {
+  const opPointTime = interpolateValue(trainSimulation[key], mToMm(opPoint.position), 'times');
+  const nextOpPointTime = interpolateValue(
+    trainSimulation[key],
+    mToMm(nextOpPoint.position),
+    'times'
+  );
+  return msToS(nextOpPointTime - opPointTime);
+}
+
+function computeMargins(
+  trainSimulation: SimulationResponseSuccess,
+  opPoint: OperationalPointWithTimeAndSpeed,
+  nextOpPoint: OperationalPointWithTimeAndSpeed,
+  selectedTrainSchedule: TrainScheduleResult,
+  pathStepId: string
+) {
+  const theoreticalMargin = getTheoreticalMargin(selectedTrainSchedule, pathStepId);
+  if (!theoreticalMargin) {
+    return emptyMargins;
+  }
+  if (!theoreticalMargin || theoreticalMargin === '0%') {
+    return emptyMargins;
+  }
```

Comment:
Since we want to display the margin in seconds even if no margin theoritical is set (but we should check if a schedule has been set or if the pathstep is a stop)
```suggestion
if ((!theoreticalMargin|| theoreticalMargin === '0%') && !pathStepIsAStop && !pathStepHasArrivalTime) {
    return emptyMargins;
  }
```

Reply:
I think we never need to return early, not anymore.

Reply:
You're right, since we don't have the via added by clicks on the map :+1: